File: schema_dump/schema_dump.py
#!/bin/python3
"""
Standalone assembler for OpenEdge database schema dumping.
This implementation directly calls OpenEdge tools without requiring the pyoe library.
"""
import os
import sys
import subprocess
import tempfile
from pathlib import Path


def dump_schema(db_path, output_file, tables="ALL", codepage="", dlc="/usr/dlc", timeout=120):
    """
    Dump the schema of an OpenEdge database to a .df file.
    
    This is a standalone implementation that directly invokes OpenEdge's
    _progres binary to run the prodict/dump_df.p procedure.
    """
    # Normalize database path (remove .db extension if present)
    db_path = Path(db_path).with_suffix("")
    
    # Check if database exists
    if not db_path.with_suffix(".db").exists():
        raise FileNotFoundError(f"Database not found: {db_path}.db")
    
    # Create output directory if needed
    output_file = Path(output_file)
    output_file.parent.mkdir(parents=True, exist_ok=True)
    
    # Get codepage from startup.pf if not specified
    if not codepage:
        pf = Path(dlc) / "startup.pf"
        try:
            for line in pf.read_text().splitlines():
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                parts = line.split()
                if len(parts) >= 2 and parts[0] == "-cpstream":
                    codepage = parts[1]
                    break
        except OSError:
            pass
    
    # ABL code to dump schema
    abl_code = f"""\
/* Standalone schema dump */
/* SESSION:PARAMETER = tables|outfile|codepage */
DEFINE VARIABLE cTables   AS CHARACTER NO-UNDO INITIAL "ALL".
DEFINE VARIABLE cOutFile  AS CHARACTER NO-UNDO.
DEFINE VARIABLE cCodePage AS CHARACTER NO-UNDO INITIAL ?.
ASSIGN
    cTables   = ENTRY(1, SESSION:PARAMETER, "|")
    cOutFile  = ENTRY(2, SESSION:PARAMETER, "|")
    cCodePage = ENTRY(3, SESSION:PARAMETER, "|") NO-ERROR.
IF cTables   = "" OR cTables   = ? THEN cTables   = "ALL".
IF cCodePage = "" THEN cCodePage = ?.
RUN prodict/dump_df.p (INPUT cTables, INPUT cOutFile, INPUT cCodePage).
"""
    
    # Write ABL code to temporary file
    with tempfile.NamedTemporaryFile(suffix=".p", mode="w", delete=False) as fh:
        fh.write(abl_code)
        proc_path = fh.name
    
    try:
        # Build command line
        progres = str(Path(dlc) / "bin" / "_progres")
        param = f"{tables}|{output_file}|{codepage}"
        cmd = [
            progres,
            "-db", str(db_path),
            "-1",      # Single-user mode
            "-b",      # Batch mode
            "-p", proc_path,
            "-param", param
        ]
        
        # Set up environment
        env = dict(os.environ)
        env["DLC"] = dlc
        
        # Add tty to PROPATH so prodict/*.r can be found
        tty = str(Path(dlc) / "tty")
        existing_propath = env.get("PROPATH", "")
        if tty not in existing_propath.split(":"):
            env["PROPATH"] = f"{tty}:{existing_propath}" if existing_propath else tty
        
        # Set terminal type
        env.setdefault("PROTERMCAP", str(Path(dlc) / "protermcap"))
        env.setdefault("TERM", "xterm")
        
        # Run OpenEdge batch process
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            env=env,
            timeout=timeout,
        )
        
        if result.returncode != 0:
            raise RuntimeError(
                f"OpenEdge batch process failed (rc={result.returncode}):\n"
                f"{result.stderr.strip()}"
            )
        
        return output_file
        
    finally:
        # Clean up temporary file
        try:
            os.unlink(proc_path)
        except OSError:
            pass


# Main execution
if __name__ == "__main__":
    db_path = sys.argv[1] if len(sys.argv) > 1 else "/var/db/test3/sports"
    df_path = sys.argv[2] if len(sys.argv) > 2 else "/tmp/sports1.df"
    print_flag = sys.argv[3] if len(sys.argv) > 3 else "false"
    
    # Dump the schema
    dump_schema(db_path, df_path)
    
    # Optionally print the output
    with open(df_path, "r") as outputfile:
        if print_flag.lower() == "true":
            print(outputfile.read())

File: schema_dump/schema_dump.c
/* 
 * C implementation of schema_dump.py for AMD64 Linux
 * Compiles with GCC
 * Usage: ./schema_dump [db_path] [df_path] [print_flag]
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <strings.h>

// Read codepage from $DLC/startup.pf
char* get_codepage_from_startup_pf(const char *dlc) {
    char pf_path[512];
    snprintf(pf_path, sizeof(pf_path), "%s/startup.pf", dlc);
    
    FILE *fp = fopen(pf_path, "r");
    if (!fp) return NULL;
    
    static char codepage[64] = {0};
    char line[512];
    
    while (fgets(line, sizeof(line), fp)) {
        // Skip empty lines and comments
        char *p = line;
        while (*p == ' ' || *p == '\t') p++;
        if (*p == '\0' || *p == '\n' || *p == '#') continue;
        
        // Look for -cpstream parameter
        if (strncmp(p, "-cpstream", 9) == 0) {
            p += 9;
            while (*p == ' ' || *p == '\t') p++;
            // Copy the codepage value
            char *end = p;
            while (*end && *end != ' ' && *end != '\t' && *end != '\n' && *end != '\r') end++;
            size_t len = end - p;
            if (len > 0 && len < sizeof(codepage)) {
                strncpy(codepage, p, len);
                codepage[len] = '\0';
                fclose(fp);
                return codepage;
            }
        }
    }
    
    fclose(fp);
    return NULL;
}

/* Return 1 if the database broker is running (lock file present), 0 otherwise */
static int db_is_running(const char *db_path) {
    char lk_path[512];
    snprintf(lk_path, sizeof(lk_path), "%s.lk", db_path);
    return (access(lk_path, F_OK) == 0);
}

#define ABL_CODE \
"DEFINE VARIABLE cTables AS CHARACTER NO-UNDO INITIAL \"ALL\".\n" \
"DEFINE VARIABLE cOutFile AS CHARACTER NO-UNDO.\n" \
"DEFINE VARIABLE cCodePage AS CHARACTER NO-UNDO INITIAL ?.\n" \
"ASSIGN cTables = ENTRY(1, SESSION:PARAMETER, \"|\")\n" \
"       cOutFile = ENTRY(2, SESSION:PARAMETER, \"|\")\n" \
"       cCodePage = ENTRY(3, SESSION:PARAMETER, \"|\") NO-ERROR.\n" \
"IF cTables = \"\" OR cTables = ? THEN cTables = \"ALL\".\n" \
"IF cCodePage = \"\" THEN cCodePage = ?.\n" \
"RUN prodict/dump_df.p (INPUT cTables, INPUT cOutFile, INPUT cCodePage).\n"

int main(int argc, char *argv[]) {
    const char *db_path = (argc > 1) ? argv[1] : "/var/db/test3/sports";
    const char *df_path = (argc > 2) ? argv[2] : "/tmp/sports1.df";
    const char *print_flag = (argc > 3) ? argv[3] : "false";
    
    // Get codepage from startup.pf
    const char *dlc = getenv("DLC");
    if (!dlc) dlc = "/usr/dlc";
    
    char *codepage = get_codepage_from_startup_pf(dlc);
    const char *cp_param = codepage ? codepage : "";

    /* Use single-user mode only when the broker is not running */
    const char *single_user = db_is_running(db_path) ? "" : "-1 ";
    
    // Create temporary file for ABL procedure
    char tmp_path[64];
    strcpy(tmp_path, "/tmp/dumpXXXXXX");
    int fd = mkstemp(tmp_path);
    if (fd < 0) {
        perror("mkstemp failed");
        return 1;
    }
    
    // Rename to add .p extension
    char final_path[128];
    snprintf(final_path, sizeof(final_path), "%s.p", tmp_path);
    rename(tmp_path, final_path);
    strcpy(tmp_path, final_path);
    
    // Write ABL code to temp file
    if (write(fd, ABL_CODE, strlen(ABL_CODE)) < 0) {
        perror("write failed");
        close(fd);
        unlink(tmp_path);
        return 1;
    }
    close(fd);
    
    // Build command to execute _progres
    char command[2048];
    snprintf(command, sizeof(command),
        "export DLC=%s PROPATH=%s/tty PROTERMCAP=%s/protermcap TERM=xterm; "
        "%s/bin/_progres -db %s %s-b -p %s -param 'ALL|%s|%s' 2>&1",
        dlc, dlc, dlc, dlc, db_path, single_user, tmp_path, df_path, cp_param);
    
    // Execute command
    int ret = system(command);
    
    // Clean up temp file
    unlink(tmp_path);
    
    // If command failed, exit
    if (ret != 0) {
        fprintf(stderr, "OpenEdge command failed with exit code %d\n", ret);
        return 1;
    }
    
    // Optionally print the output file
    if (strcasecmp(print_flag, "true") == 0) {
        FILE *fp = fopen(df_path, "r");
        if (fp) {
            char buffer[8192];
            size_t n;
            while ((n = fread(buffer, 1, sizeof(buffer) - 1, fp)) > 0) {
                buffer[n] = '\0';
                printf("%s", buffer);
            }
            fclose(fp);
        }
    }
    
    return 0;
}
